waspc/src/Wasp/Generator/SdkGenerator/Server/OAuthG.hs

```diff
@@ -0,0 +1,97 @@
+import StrongPath (File', Path', Rel, reldir, relfile, (</>))
+import qualified StrongPath as SP
+import Wasp.AppSpec (AppSpec)
+import qualified Wasp.AppSpec.App as AS.App
+import qualified Wasp.AppSpec.App.Auth as AS.App.Auth
+import qualified Wasp.AppSpec.App.Auth as AS.Auth
+import qualified Wasp.AppSpec.App.Dependency as AS.Dependency
+import qualified Wasp.AppSpec.Valid as AS.Valid
+import Wasp.Generator.AuthProviders (discordAuthProvider, getEnabledAuthProvidersJson, gitHubAuthProvider, googleAuthProvider, keycloakAuthProvider)
+import Wasp.Generator.AuthProviders.OAuth
+  ( OAuthAuthProvider,
+    clientOAuthCallbackPath,
+    serverExchangeCodeForTokenHandlerPath,
+    serverOAuthCallbackHandlerPath,
+    serverOAuthLoginHandlerPath,
+  )
+import qualified Wasp.Generator.AuthProviders.OAuth as OAuth
+import Wasp.Generator.FileDraft (FileDraft)
+import Wasp.Generator.Monad (Generator)
+import qualified Wasp.Generator.SdkGenerator.Common as C
+import Wasp.Util ((<++>))
+
+genOAuth :: AS.Auth.Auth -> Generator [FileDraft]
+genOAuth auth
+  | AS.Auth.isExternalAuthEnabled auth =
+      sequence
+        [ genIndexTs auth,
+          genRedirectHelper,
+          genFileCopy [relfile|server/oauth/env.ts|],
+          genFileCopy [relfile|server/oauth/oneTimeCode.ts|]
+        ]
+        <++> genOAuthProvider discordAuthProvider (AS.Auth.discord . AS.Auth.methods $ auth)
+        <++> genOAuthProvider googleAuthProvider (AS.Auth.google . AS.Auth.methods $ auth)
+        <++> genOAuthProvider keycloakAuthProvider (AS.Auth.keycloak . AS.Auth.methods $ auth)
+        <++> genOAuthProvider gitHubAuthProvider (AS.Auth.gitHub . AS.Auth.methods $ auth)
+  | otherwise = return []
+  where
+    genFileCopy = return . C.mkTmplFd
+
+genIndexTs :: AS.Auth.Auth -> Generator FileDraft
+genIndexTs auth = return $ C.mkTmplFdWithData tmplFile tmplData
+  where
+    tmplFile = [relfile|server/oauth/index.ts|]
+    tmplData =
+      object
+        [ "enabledProviders" .= getEnabledAuthProvidersJson auth
+        ]
+
+genRedirectHelper :: Generator FileDraft
+genRedirectHelper = return $ C.mkTmplFdWithData [relfile|server/oauth/redirect.ts|] tmplData
+  where
+    tmplData =
+      object
+        [ "serverOAuthCallbackHandlerPath" .= serverOAuthCallbackHandlerPath,
+          "clientOAuthCallbackPath" .= clientOAuthCallbackPath,
+          "serverOAuthLoginHandlerPath" .= serverOAuthLoginHandlerPath,
+          "serverExchangeCodeForTokenHandlerPath" .= serverExchangeCodeForTokenHandlerPath
+        ]
+
+genOAuthProvider ::
+  OAuthAuthProvider ->
+  Maybe AS.Auth.ExternalAuthConfig ->
+  Generator [FileDraft]
+genOAuthProvider provider maybeUserConfig
+  | isJust maybeUserConfig = sequence [genOAuthConfig provider]
+  | otherwise = return []
+
+genOAuthConfig ::
+  OAuthAuthProvider ->
+  Generator FileDraft
+genOAuthConfig provider = return $ C.mkTmplFdWithData tmplFile tmplData
+  where
+    tmplFile = [reldir|server/oauth/providers|] </> providerTsFile
```

Comment:
I suggest extracting the root folder to avoid repeating it in all generator functions.

Reply:
I now see that no other module did this, so thanks for taking a step beyond :)